File: 0134-gas-station/0134-gas-station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        diffs = []
        total_cost = 0

        for i in range(len(gas)):
            diffs.append(gas[i] - cost[i])
            total_cost += diffs[i]

        if (total_cost < 0):
            return -1

        start_idx = 0
        total_cost = 0
        for i in range (len(diffs)):
            total_cost += diffs[i]
            if (total_cost < 0):
                start_idx = i + 1
                total_cost = 0

        return start_idx

File: 0134-gas-station/0134-gas-station.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int n = gas.length;
        
        int[] diff = new int[n];
        int totalDiff = 0;
        for (int i = 0; i<n; i++){
            diff[i] = gas[i] - cost[i];
            totalDiff += diff[i];
        }

        if (totalDiff < 0) {return -1;}

        int startIdx = -1;
        totalDiff = 0;
        for (int i = 0; i<n; i++){
            totalDiff += diff[i];

            if (totalDiff < 0){ 
                startIdx = -1;
                totalDiff = 0;
            } else if (startIdx == -1){
                startIdx = i;
            }
        }
        // System.out.println(Arrays.toString(diff));
        // System.out.println(totalDiff);

        return startIdx; 
    }
}